From 4633f3395e5701ab0e746e638a1cfe91f3ddb6d2 Mon Sep 17 00:00:00 2001
From: Oleh Nykyforchyn <olen.nyk@gmail.com>
Date: Mon, 11 Aug 2025 22:49:51 +0300
Subject: [PATCH 3/4] xserver: hw/xfree86/config: install 10-nvidia.conf

It is patch 3/4 of a series that provides a convenient way to specify
IgnoreABI on a per-driver basis.

This patch installs a /usr/share/X11/xorg.conf.d/10-nvidia.conf file with
an OutputClass that makes the loader ignore ABI for all devices managed by
the nvidia driver.

Signed-off-by: Oleh Nykyforchyn <oleh.nyk@gmail.com>
---
 config/10-nvidia.conf.in | 16 ++++++++++++++++
 config/meson.build       |  8 ++++++++
 2 files changed, 24 insertions(+)
 create mode 100644 config/10-nvidia.conf.in

diff --git a/config/10-nvidia.conf.in b/config/10-nvidia.conf.in
new file mode 100644
index 000000000..d4ca7c2a5
--- /dev/null
+++ b/config/10-nvidia.conf.in
@@ -0,0 +1,16 @@
+# This xorg.conf.d configuration snippet configures the X server to
+# automatically load the nvidia X driver when it detects a device driven by the
+# nvidia-drm.ko kernel module and ignores nvidia_drv.so ABI version.
+# Please note that this works on Linux kernels version 6.12 or higher
+# with CONFIG_DRM enabled, and only if the nvidia-drm.ko kernel module
+# is loaded before the X server is started.
+
+Section "OutputClass"
+	Identifier "NVidia"
+	MatchDriver "nvidia-drm"
+	Driver "nvidia"
+	Option "AllowEmptyInitialConfiguration"
+	#Add this to ServerFlags if your xserver does not recognize it in OutputClass
+	Option "IgnoreABI"
+	ModulePath "@libdir@/nvidia/xorg"
+EndSection
diff --git a/config/meson.build b/config/meson.build
index 15f21a053..a4f73c7cd 100644
--- a/config/meson.build
+++ b/config/meson.build
@@ -26,6 +26,14 @@ endif
 if build_xorg
     install_data('10-quirks.conf',
                  install_dir: join_paths(get_option('datadir'), 'X11/xorg.conf.d'))
+    nvidia_conf = configuration_data()
+    nvidia_conf.set('libdir', join_paths(get_option('prefix'),get_option('libdir')))
+    configure_file(
+        input: '10-nvidia.conf.in',
+        output: '10-nvidia.conf',
+        configuration: nvidia_conf,
+        install_dir: join_paths(get_option('datadir'), 'X11/xorg.conf.d'),
+    )
 endif
 
 libxserver_config = static_library('xserver_config',
-- 
2.33.0

